src/wayland/tablet/tablet_tool.rs

```diff
@@ -0,0 +1,398 @@
+use std::ops::Deref as _;
+use std::sync::Mutex;
+use std::{cell::RefCell, rc::Rc};
+
+use crate::backend::input::{TabletToolDescriptor, TabletToolType};
+use crate::wayland::seat::{CursorImageAttributes, CursorImageStatus};
+use wayland_protocols::unstable::tablet::v2::server::{
+    zwp_tablet_seat_v2::ZwpTabletSeatV2,
+    zwp_tablet_tool_v2::{self, ZwpTabletToolV2},
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::Filter;
+
+use crate::wayland::{compositor, SERIAL_COUNTER};
+
+use super::tablet::TabletHandle;
+
+static CURSOR_IMAGE_ROLE: &str = "cursor_image";
+
+#[derive(Debug, Default)]
+struct TabletTool {
+    instances: Vec<ZwpTabletToolV2>,
+    focus: Option<WlSurface>,
+
+    is_down: bool,
+
+    pending_pressure: Option<f64>,
+    pending_distance: Option<f64>,
+    pending_tilt: Option<(f64, f64)>,
+    pending_slider: Option<f64>,
+    pending_rotation: Option<f64>,
+}
+
+impl TabletTool {
+    fn proximity_in(
+        &mut self,
+        (x, y): (f64, f64),
+        (focus, (sx, sy)): (WlSurface, (f64, f64)),
+        tablet: &TabletHandle,
+        time: u32,
+    ) {
+        let wl_tool = self
+            .instances
+            .iter()
+            .find(|i| i.as_ref().same_client_as(focus.as_ref()));
+
+        if let Some(wl_tool) = wl_tool {
+            tablet.with_focused_tablet(&focus, |wl_tablet| {
+                let serial = SERIAL_COUNTER.next_serial();
```

Comment:
Such methods should take the serial as argument rather than increment it themselves. The serial can then be used by the client to request some action in response to, so the compositor needs to be aware of them, meaning it should provide them.

Reply:
The main reason for that, was because `proximity_in()` can be called by `motion()` in some cases, but in most cases `motion` does not use serial for anything, so... would it be ok to simply do nothing with supplied serials in those cases?

Reply:
Yes that would be ok. This is more or less what we are already doing with pointer events.